libbpf-tools/fsslower.bpf.c

```diff
@@ -0,0 +1,202 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+/* Copyright (c) 2021 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "bits.bpf.h"
+#include "fsslower.h"
+
+#define MAX_ENTRIES	10240
+
+const volatile pid_t target_pid = 0;
+const volatile __u64 min_lat_ms = 0;
+
+struct data {
+	__u64 ts;
+	loff_t start;
+	loff_t end;
+	struct file *fp;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct data);
+} starts SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+static int probe_entry(struct file *fp, loff_t start, loff_t end)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	struct data data = {};
+
+	if (!fp)
+		return 0;
+
+	if (target_pid && target_pid != pid)
+		return 0;
+
+	data.ts = bpf_ktime_get_ns();
+	data.fp = fp;
+	data.start = start;
+	data.end = end;
+	bpf_map_update_elem(&starts, &tid, &data, BPF_ANY);
+	return 0;
+}
+
+static int probe_exit(void *ctx, enum fs_file_op op, ssize_t size)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	__u64 end_ns, delta_us;
+	const __u8 *file_name;
+	struct data *datap;
+	struct event event = {};
+	struct dentry *dentry;
+	struct file *fp;
+
+	datap = bpf_map_lookup_elem(&starts, &tid);
+	if (!datap)
+		return 0;
+
+	bpf_map_delete_elem(&starts, &tid);
+
+	end_ns = bpf_ktime_get_ns();
+	delta_us = (end_ns - datap->ts) / 1000;
+	if (delta_us <= min_lat_ms * 1000)
```

Comment:
instead of having min_lat_ms in milliseconds and do nanosecond to millisecond convertion, why not use nanosecond units for min_lat and do a simple and fast comparison?